### PR TITLE
[CIR] Upstream initial attribute support

### PR DESCRIPTION
**Pull request by dkolsen-pgi**
Upstream several ClangIR-specific MLIR attributes, in particular attributes for integer, floating-point, and null pointer constants. These are the first ClangIR attributes to be upstreamed, so infrastructure changes are included, such as the table-gen file `CIRAttrs.td`.

Attributes can be used as the initial values for global variables.  The existing automated test global-var-simple.cpp includes initial values for some of the global variables in the test.

### CODE REVIEW

File: clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h
Comment (erichkeane) [inline clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h:34]: Could we do a better variable name than `v`?  I presume it means `value`?  But that is really only something I can guess at.

File: clang/lib/CIR/Dialect/IR/CIRAttrs.cpp
Comment (erichkeane) [inline clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:62]: Agreed with this TODO.  The textual representation of 'null' is a little fragile for me.

File: clang/lib/CIR/Dialect/IR/CIRAttrs.cpp
Comment (erichkeane) [inline clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:101]: Do these need to `return`?  It seems that continuing with a garbage `value` here could be surprising?

**Review by erichkeane — COMMENTED** [review]
Few small things, else LGTM.  I'm away until next Thursday, so the others can approved if they are around.

File: clang/lib/CIR/Dialect/IR/CIRAttrs.cpp
Comment (erichkeane) [reply clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:62]: ^^ Nothing to do in this patch, just highlighting this TODO for posterity.

File: clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h
Comment (dkolsen-pgi) [reply clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h]: Done.

I am still calibrating when to use the same names that are in the ClangIR incubator and when to choose better names.

File: clang/lib/CIR/Dialect/IR/CIRAttrs.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:101]: I don't want to add a return here because the code should fall through and try to parse the closing `>` character.  That increases the chance that the parser will recover from an invalid integer value and not get completely lost.  But I did rearrange the code some so that an uninitialized garbage value is never returned.

Comment (erichkeane) [reply clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:101]: Makes sense, this LGTM.

**Review by erichkeane — APPROVED** [review]
Looks ok to me.

**Pull request merged**
